internal/batches/executor/partial_eval.go

```diff
@@ -0,0 +1,388 @@
+			isStatic = false
+			break
+		}
+	}
+	if !isStatic {
+		return isStatic, false, nil
+	}
+
+	return true, isTrueOutput(t.Tree.Root), nil
+}
+
+// parseAndPartialEval parses input as a text/template and then attempts to
+// partially evaluate the parts of the template it can evaluate ahead of time
+// (meaning: before we've executed any batch spec steps and have a full
+// StepContet available).
+//
+// If it's possible to evaluate a parse.ActionNode (which is what sits between
+// delimiters in a text/template), the node is rewritten into a parse.TextNode,
+// to make it look like it's always been text in the template.
+//
+// Partial evaluation is done in a best effort manner: if it's not possible to
+// evaluate a node (because it requires information that we only later get, or
+// because it's too complex, etc.) we degrade gracefully and simply abort the
+// partial evaluation and leave the node as is.
+//
+// It also should be noted that we don't do "full" partial evaluation: if we
+// come across value that we can't partially evaluate we abort the process *for
+// the whole node* without replacing the sub-nodes that we've successfully
+// evaluated. Why? Because we can't construct correct `*parse.Node` from
+// outside the `parse` package. In other words: we evaluate
+// all-parse.ActionNode-or-nothing.
+func parseAndPartialEval(input string, ctx *StepContext) (*template.Template, error) {
+	t, err := template.
+		New("partial-eval").
+		Delims(startDelim, endDelim).
+		Funcs(builtins).
+		Funcs(ctx.ToFuncMap()).
+		Parse(input)
+
+	if err != nil {
+		return nil, err
+	}
+
+	for i, n := range t.Tree.Root.Nodes {
+		t.Tree.Root.Nodes[i] = rewriteNode(n, ctx)
+	}
+
+	return t, nil
+}
+
+// rewriteNode takes the given parse.Parse and tries to partially evaluate it.
+// If that's possible, the output of the evaluation is turned into text and
+// instead of the node that was passed in a new parse.TextNode is returned that
+// represents the output of the evaluation.
+func rewriteNode(n parse.Node, ctx *StepContext) parse.Node {
+	switch n := n.(type) {
+	case *parse.ActionNode:
+		if val, ok := evalPipe(ctx, n.Pipe); ok {
+			var out bytes.Buffer
+			fmt.Fprint(&out, val.Interface())
+			return &parse.TextNode{
+				Text:     out.Bytes(),
+				Pos:      n.Pos,
+				NodeType: parse.NodeText,
+			}
+		}
+
+		return n
+
+	default:
+		return n
+	}
+}
+
+// noValue is returned by the functions that partially evaluate a parse.Node
+// to signify that evaluation was not possible or did not yield a value.
+var noValue reflect.Value
+
+func evalPipe(ctx *StepContext, p *parse.PipeNode) (reflect.Value, bool) {
+	// If the pipe contains declaration we abort evaluation.
+	if len(p.Decl) > 0 {
+		return noValue, false
+	}
+
+	// TODO: Support finalVal
+	var (
+		// finalVal is the value of the previous Cmd in a pipe (i.e. `${{ 3 + 3 | eq 6 }}`)
+		// It needs to be the final (fixed) argument of a call if it's set.
+		finalVal reflect.Value
+
+		ok bool
+	)
```

Comment:
Maybe these could be named return arguments instead?

Reply:
Ah! That's good. Fixed.